src/sinks/util/service2.rs

```diff
@@ -174,6 +188,61 @@ where
     }
 }
 
+// === map ===
+
+pub struct MapLayer<R1, R2> {
+    f: Arc<dyn Fn(R1) -> R2 + Send + Sync + 'static>,
+}
+
+impl<S, R1, R2> Layer<S> for MapLayer<R1, R2>
+where
+    S: Service<R2>,
+{
+    type Service = Map<S, R1, R2>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Map {
+            f: self.f.clone(),
+            inner,
+        }
+    }
+}
+
+pub struct Map<S, R1, R2> {
+    f: Arc<dyn Fn(R1) -> R2 + Send + Sync + 'static>,
+    inner: S,
+}
+
+impl<S, R1, R2> Service<R1> for Map<S, R1, R2>
+where
+    S: Service<R2>,
+    crate::Error: From<S::Error>,
+{
+    type Response = S::Response;
+    type Error = crate::Error;
+    type Future = futures::future::MapErr<S::Future, fn(S::Error) -> crate::Error>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner
+            .poll_ready(cx)
+            .map(|result| result.map_err(|e| e.into()))
+    }
+
+    fn call(&mut self, req: R1) -> Self::Future {
+        let req = (self.f)(req);
+        self.inner.call(req).map_err(|e| e.into())
+    }
+}
+
+impl<S: Clone, R1, R2> Clone for Map<S, R1, R2> {
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
```

Comment:
I don't see this code actually being used elsewhere in this PR. Did I miss something? It looks like it could be useful, but if it's not used it should get removed.

Reply:
It's used here: https://github.com/timberio/vector/blob/ee86731a755abe3c5d119f78dd89bae0eaaca6e8/src/sinks/gcp/cloud_storage.rs#L212

Reply:
Ha, interesting. That didn't show up in the diff so I didn't realize it was used. I see now this `Map` is a feature that wasn't initially carried over from the older `service.rs` module. :+1: